src/core/SampleBufferCache.cpp

```diff
@@ -0,0 +1,80 @@
+/*
+ * SampleBufferCache.cpp - Cache for storing SampleBuffer's
+ *
+ * Copyright (c) 2024 saker <sakertooth@gmail.com>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "SampleBufferCache.h"
+
+#include <cassert>
+#include <cerrno>
+#include <stdexcept>
+
+namespace lmms {
+
+SampleBufferCache::SampleBufferCache(size_t capacity)
+	: m_capacity(capacity)
+{
+	if (capacity == 0) { throw std::runtime_error{"Capacity of cache cannot be zero"}; }
+}
+
+auto SampleBufferCache::add(const QString& key, Buffer buffer) -> bool
+{
+	if (buffer == nullptr || m_cache.contains(key)) { return false; }
```

Comment:
I just realized this breaks all my tests since they use empty buffers.

I guess I will have to delete my tests, but I think there should be a better way. Maybe I will add the caching support directly within the `SampleLoader` create functions.

I don't think an empty buffer should signify a buffer that failed to load, even if we do return one in that case.

Reply:
Maybe you should make the `SampleLoader::createBufferFrom*` methods return nullptr on failure instead of `SampleBuffer::emptyBuffer()`. There doesn't seem to be much that you'd have to change in calling code.

`Sample`'s constructor from a `SampleBuffer` would need to check if it's null and create a sample buffer if so:

```c++
Sample::Sample(std::shared_ptr<const SampleBuffer> buffer)
	: m_buffer(buffer ? std::move(buffer) : std::make_shared<const SampleBuffer>())
	, m_startFrame(0)
	, m_endFrame(m_buffer->size())
	, m_loopStartFrame(0)
	, m_loopEndFrame(m_buffer->size())
{
}
```

Also, there might be a bug in `void SlicerT::updateFile(QString file)` when you move the `std::shared_ptr` from `SampleBuffer::emptyBuffer()`. It could leave the global empty buffer in a moved-from state.

With the changes I suggested above and this potential way to easily misuse the empty buffer, you might be better off removing `SampleBuffer::emptyBuffer()` entirely.

Reply:
Now that you are returning `nullptr` from `SampleLoader::createBufferFromFile()` when it fails (which is great, imo), I think you need to take a look at the call sites to make sure they are handling `nullptr` correctly. If `Sample` assumes its sample buffer is never null, you'll need to fix the `Sample::Sample(std::shared_ptr<const SampleBuffer> buffer)` constructor like I mentioned in my comment above.

Reply:
Is there a reason we don't use optional here? Then it would be a conpile-time error to not handle the empty case at call sites.

Reply:
I think it would be better for `SampleCache::get` to just propagate the exception up near some GUI boundary, so that it can display the message box there.

Exceptions provide more detail to an error for the user instead of just returning `nullptr`.

Reply:
This is one of those moments I wish we had `std::expected`